src/cert/der_test.rs

```diff
@@ -0,0 +1,333 @@
+    })
+}
+
+#[test]
+fn uint_9000() -> Result {
+    testdata::NINE_THOUSAND.read_all(BadEncoding, |buf| {
+        let int = der::uint(buf)?;
+        assert_eq!(int.as_slice_less_safe(), [0x23, 0x28]);
+        Ok(())
+    })
+}
+
+#[test]
+fn uint_huge() -> Result {
+    testdata::HUGE_INT.read_all(BadEncoding, |buf| {
+        let int = der::uint(buf)?;
+        assert_eq!(int.as_slice_less_safe(), [0x55; 5]);
+        Ok(())
+    })
+}
+
+#[test]
+fn uint_00() {
+    let mut reader = Reader::new(testdata::DOUBLE_ZERO);
+    assert!(der::uint(&mut reader).is_err());
+}
+
+#[test]
+fn uint_neg() {
+    let mut reader = Reader::new(testdata::NEGATIVE);
+    assert!(der::uint(&mut reader).is_err());
+}
+
+#[test]
+fn u32_0() -> Result {
+    testdata::ZERO.read_all(BadEncoding, |buf| {
+        assert_eq!(der::u32(buf)?, 0);
+        Ok(())
+    })
+}
+
+#[test]
+fn u32_42() -> Result {
+    testdata::FORTY_TWO.read_all(BadEncoding, |buf| {
+        assert_eq!(der::u32(buf)?, 42);
+        Ok(())
+    })
+}
+
+#[test]
+fn u32_128() -> Result {
+    testdata::ONE_TWENTY_EIGHT.read_all(BadEncoding, |buf| {
+        assert_eq!(der::u32(buf)?, 128);
+        Ok(())
+    })
+}
+
+#[test]
+fn u32_9000() -> Result {
+    testdata::NINE_THOUSAND.read_all(BadEncoding, |buf| {
+        assert_eq!(der::u32(buf)?, 9000);
+        Ok(())
+    })
+}
+
+#[test]
+fn u32_huge() {
+    let mut reader = Reader::new(testdata::HUGE_INT);
+    assert!(der::u32(&mut reader).is_err());
+}
+
+#[test]
+fn u32_00() {
+    let mut reader = Reader::new(testdata::DOUBLE_ZERO);
+    assert!(der::u32(&mut reader).is_err());
+}
+
+#[test]
+fn u32_neg() {
+    let mut reader = Reader::new(testdata::NEGATIVE);
+    assert!(der::u32(&mut reader).is_err());
+}
+
+#[test]
+fn null() -> Result {
+    testdata::NULL.read_all(BadEncoding, der::null)
+}
+
+#[test]
+fn null_bad_tag() {
+    let mut reader = Reader::new(testdata::FORTY_TWO);
+    assert!(der::null(&mut reader).is_err());
+}
+
+#[test]
+fn null_nonempty() {
+    let mut reader = Reader::new(testdata::NONEMPTY_NULL);
+    assert!(der::null(&mut reader).is_err());
+}
+
+#[test]
+fn opt_bool_true() -> Result {
+    testdata::TRUE.read_all(BadEncoding, |buf| {
+        assert_eq!(der::opt_bool(buf)?, Some(true));
+        Ok(())
+    })
+}
+
+#[test]
+fn opt_bool_false() -> Result {
+    testdata::FALSE.read_all(BadEncoding, |buf| {
+        assert_eq!(der::opt_bool(buf)?, Some(false));
+        Ok(())
+    })
+}
+#[test]
+fn opt_bool_wrong_tag() -> Result {
+    testdata::FORTY_TWO.read_all(BadEncoding, |buf| {
+        assert!(der::opt_bool(buf)?.is_none());
+        buf.skip_to_end();
+        Ok(())
+    })
+}
+
+#[test]
+fn opt_bool_empty() -> Result {
+    testdata::EMPTY.read_all(BadEncoding, |buf| {
+        assert!(der::opt_bool(buf)?.is_none());
+        Ok(())
+    })
+}
+
+#[test]
+fn opt_bool_bad() {
+    let mut reader = Reader::new(testdata::BAD_BOOL);
+    assert!(der::opt_bool(&mut reader).is_err());
+}
```

Comment:
Does it make sense to have tests for the `oid!` macro?

Reply:
Not particularly. Its sole purpose in life is generating constants that der-ascii can emit, too.

(Also, I may wind up getting rid of `oid!` and just using der-ascii to generate the OIDs. we'll see.